### ISSUE TITLE
Draft: Where should our different kinds of data be stored?

### ISSUE THREAD

**Issue by turbomam** [microbiomedata/issues#630]
This morning I am feeling very inspired by some work that @anastasiyaprymolenna , @sierra-moxon and others were doing last night. I'm talking about reviewing a [PR I made for simplifying ChemicalConversionProcess](https://github.com/microbiomedata/berkeley-schema-fy24/pull/100), especially identifying where reference knowledge about chemical substances should live.

For an `Instrument` indexed by it's serial number, reference knowledge might be the vendor name, the model number, and/ or the smallest and largest entities it can detect in a given configuration. For a `Substance`, the reference knowledge would include the SMILES and InCHI representations, molecular weight, etc. Those resources can be used in different ways, but some things hold constant for them  no matter how they are used.

_This issue is currently a draft. I will be adding clarification and examples. That doesn't mean you can't comment on it now or ask questions._

An initial assumption: we will not allow totally unconstrained specifications of `Instrument`s, `Substance`s, organizations, etc. That means the primary key for those instances won't be a string slot, or even an unconstrained CURIe slot.

## Some storage options:
1. Put **all** knowledge about a thing in in-lined instances (like `Substance`s) that are embedded in some other instances (like `ChemicalConversionProcess`es) which have an identifier and can be found in a MongoDB collection. This is the most verbose and free-form solution. It will almost certainly lead to duplicate and /or contradictory information.
2. Put all reference knowledge about something into annotations on permissible values in an enumeration, then mention that permissible value in classes with an identifier. That class could also reflect the way that some thing (as specified by a permissible value) was used, like the concentration, volume or mass of a substance. This solution requires searches to traverse both the data in MongoDB and the enumeration(s) in the schema.
3. Save the reference knowledge as instances of some class in MongoDB. It would be a new collection like `substance_set`. berkeley-schema-fy24 is already moving in that direction for [Instrument](https://microbiomedata.github.io/berkeley-schema-fy24/Instrument/)s. In this case, searches would only need to traverse MongoDB
4. Save the reference knowledge in the nmdc-ontology. One justification for this is that there doesn't appear to be any term, in any of the ontologies or databases that we prioritize, that uniquely identifies de-ionized water. We could create a term **in** nmdco ,at least as a stopgap, and then shop it around to other ontologies we work with.

## The crux issue:

We need to support searches in which the user uses general terms like "find LC/MS proteomics data where the proteins were digested in the presence of the detergent tween". We know that tween comes in several formulations with different characteristics (and may shouldn't ever be put anywhere near a LC/MS system!). So we may be tempted to say that we must have multiple entities in our system (using one of the storage solutions above) that all use tween as the primary key because that's what the user will query for.

Of course your can't have multiple identical primary keys in a data storage system, and that may be seems as justification to bundle the reference data (like InChI specifications) in the in-lined instances, option 1.

I am taking the position that reference knowledge should be separated from usage instances, and that the reference knowledge records must use distinct, un-ambiguous identifiers, If several of these entities share trivial names, then those should be aliases. At this point it becomes the responsibility of the user interface (a website, API, etc.) to support searches that account for ambiguity. If the reference knowledge is stored intelligently, then the development of search tools shouldn't be too difficult.

----

Note that I am not addressing "workflow data" here, like FASTQ inputs or GFF outputs. I would like to see or create a document that explains where these live (mostly as files on NERSC?) and how to tie them into schema-compliant records in MongDB (DataObject URLs?)

**Comment by anastasiyaprymolenna** [comment]
Related to ChemicalConversionProcess refactoring https://github.com/microbiomedata/nmdc-schema/issues/1842

**Comment by turbomam** [comment]
> Related to ChemicalConversionProcess refactoring [microbiomedata/nmdc-schema#1842](https://github.com/microbiomedata/nmdc-schema/issues/1842)

Yes, among other modeling patterns that involve reference and usage information

**Comment by mslarae13** [comment]
@turbomam 
This related to ChemicalConversionProcess. seems like the remainder of this issue is a "thought experiment" .

I think the requirement for Metabolink2 is done. Will remove from squad board

**Comment by mslarae13** [comment]
Won't be resolved as part of the metabolink 2 squad. Moving this out of the task.

**Comment by mslarae13** [comment]
@turbomam to read through and review / closed
if the issue has to do with chemical used or instruments we've resolved this.
@kheal generating metadata for material processing is working well as is.

**Comment by turbomam** [comment]
from Claude

1. Instruments

Currently uses Pattern 3 (Reference collection) + Pattern 2 (Enum annotations)

- Class: Instrument (basic_classes.yaml:117-138)
  - Has id, vendor, model slots
  - Unique key: vendor + model combination
  - ID pattern: nmdc:inst-{shoulder}-{blade}
- Collection: instrument_set in Database (nmdc.yaml:1312-1316)
- Slot reference: instrument_used (basic_classes.yaml:671-677)
  - range: Instrument - references the class, not inlined
  - Used in DataGeneration and MaterialProcessing
- Enums for properties:
  - InstrumentModelEnum (basic_classes.yaml:829-928+) - has aliases, meanings, comments, structured_aliases
  - InstrumentVendorEnum (basic_classes.yaml:1077+) - has meanings, aliases

Verdict: Instruments already follow your recommended approach - reference knowledge in a collection (instrument_set), with enum annotations for controlled vocabulary of vendors/models.

2. Substances (Chemicals)

Currently uses Pattern 2 (Enum annotations) + Pattern 1 (Inlined instances)

- Enum: ChemicalEntityEnum (core.yaml:144-232)
  - 25+ chemicals with meaning (CHEBI/MS mappings) and comments
  - Example: trypsin: meaning: MS:1001251
- Inlined class: PortionOfSubstance (core.yaml:1220-1233)
  - Has known_as (range: ChemicalEntityEnum), mass, volume, final_concentration, substance_role
  - No identifier slot - not a NamedThing, cannot be stored independently
  - Used inlined in substances_used slot of ChemicalConversionProcess
- Deprecated: ChemicalEntity class (deprecated.yaml:95-120) - was going to be Pattern 3 but deprecated
- No collection: There is no substance_set or chemical_entity_set in Database

Verdict: Substances currently use Pattern 2 (enum annotations for reference knowledge) + Pattern 1 (inlined PortionOfSubstance for usage knowledge). This is exactly the scenario your
issue was addressing - reference knowledge is embedded in enum annotations, not in a queryable collection.

3. Organizations

Currently uses Pattern 2 (Enum annotations) only

- Enum: ProcessingInstitutionEnum (basic_slots.yaml:1090-1120)
  - Has meaning (ROR identifiers), title, aliases
  - Example: EMSL: meaning: ror:04rc0xn13
- Slot: processing_institution (basic_slots.yaml:351-353)
  - range: ProcessingInstitutionEnum - enum value, not a class reference
- Related: principal_investigator (basic_classes.yaml:568-572)
  - range: PersonValue - a different approach (inlined PersonValue)
- No class: There is no Organization class
- No collection: There is no organization_set in Database

Verdict: Organizations only use Pattern 2. Reference knowledge (ROR IDs, full names) is stored in enum annotations. There's no way to query organizations independently or attach
additional metadata.


Summary Table

| Concern       | Pattern 1 (Inlined)      | Pattern 2 (Enum annotations)    | Pattern 3 (Collection) | Pattern 4 (Ontology) |
|---------------|--------------------------|---------------------------------|------------------------|----------------------|
| Instruments   | No                       | Yes (model/vendor enums)        | Yes (instrument_set)   | No                   |
| Substances    | Yes (PortionOfSubstance) | Yes (ChemicalEntityEnum)        | No                     | No                   |
| Organizations | No                       | Yes (ProcessingInstitutionEnum) | No                     | No                   |

Instruments are the only one of the three using your recommended Pattern 3 (collection-based reference data). Substances and Organizations still rely primarily on Pattern 2 (enum annotations).

**Comment by turbomam** [comment]
The in-lining in `PortionOfSubstance` is really usage (volume, concentration, etc), not reference (formula, molecular weight). 

The reference data is captured in `known_as` -> `ChemicalEntityEnum`

I anticipate that we can migrate to more use of Pattern 4 (Ontology) when we all become more familiar with @sierra-moxon's ontology loader
